caffe2/operators/onnx_while_op.h

```diff
@@ -0,0 +1,233 @@
+    *iteration_var_ptr = 0ll;
+
+    // Input condition var. This requires special handling
+    input_condition_var_->Resize(1);
+    auto* input_condition_var_ptr =
+        input_condition_var_->template mutable_data<bool>();
+    *input_condition_var_ptr = first_iter_condition;
+
+    // Output condition var. This is yielded by the body net and we will use its
+    // value to determine further iteration
+
+    condition_var_->Resize(1);
+    auto* condition_var_ptr = condition_var_->template mutable_data<bool>();
+
+    auto valid_iter_num = [this, max_trip_count](int64_t i) {
+      if (has_trip_count_) {
+        return i < max_trip_count;
+      } else {
+        return true;
+      }
+    };
+
+    auto condition_true =
+        [this, first_iter_condition, condition_var_ptr](int64_t i) {
+          if (has_cond_) {
+            if (i == 0) {
+              return (bool)first_iter_condition;
+            } else {
+              return (bool)*condition_var_ptr;
+            }
+          } else {
+            return true;
+          }
+        };
+
+    // Allocate scan_outputs for zero-iteration case
+    for (int i = 0; i < num_scan_outputs; ++i) {
+      Output(i + num_loop_carried_deps)->Resize(0);
+      Output(i + num_loop_carried_deps)->template mutable_data<int32_t>();
+    }
+
+    // Use this to keep track of the sizes of the scan outputs and validate
+    // they're the same across iterations.
+    std::vector<std::vector<TIndex>> scan_outputs_sizes;
+
+    while (true) {
+      int64_t itr = *iteration_var_ptr;
+      if (valid_iter_num(itr) && condition_true(itr)) {
+        if (!body_net_->Run()) {
+          return false;
+        }
+        // Copy forward loop-carried dependencies
+        for (int i = 0; i < num_loop_carried_deps; ++i) {
+          Blob* b = ws_->GetBlob(body_net_->external_output()[i + 1]);
+          const Tensor<Context>& t = b->template Get<Tensor<Context>>();
+          lcd_tensors_[i]->CopyFrom(t);
+        }
+        // Copy out scan_outputs
+        for (int i = 0; i < num_scan_outputs; ++i) {
+          int net_output_idx = i + 1 + num_loop_carried_deps;
+          const Tensor<Context>& scan_output =
+              ws_->GetBlob(body_net_->external_output()[net_output_idx])
+                  ->template Get<Tensor<Context>>();
+          auto* scan_output_target = Output(i + num_loop_carried_deps);
+          if (itr == 0) {
+            auto dims = scan_output.dims();
+            scan_outputs_sizes.push_back(dims);
+            dims.insert(dims.begin(), 1);
+            scan_output_target->Resize(dims);
+            scan_output_target->CopyFrom(scan_output);
+          } else {
+            auto dims = scan_output.dims();
+            CAFFE_ENFORCE_EQ(
+                dims,
+                scan_outputs_sizes[i],
+                "Size of scan output changed across iterations");
+            dims.insert(dims.begin(), itr);
+            scan_output_target->Extend(1, 2.0f, &context_);
```

Comment:
2.0f is "resize percentage" here. I assume you wanted to do 2x resize here, like vector does? Then it should be 100